{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ed042d15-4b09-41f9-99b4-608b3452e799",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Use this notebook to generate pgBoost SNP-gene linking features ##\n",
    "\n",
    "## DEFINE RELEVANT PARAMETERS\n",
    "\n",
    "# Output folder: define a folder to send output (feature matrix, etc.)\n",
    "output_folder = \".\" ## Change to desired folder name\n",
    "\n",
    "# INPUT FILE PATHS #\n",
    "# For each dataset/cell type (for which multiome scores are being used as pgBoost input, e.g. K562 cells from dataset 1), provide:\n",
    "# File paths to linking scores for each constituent method in its respective list (Signac, SCENT, Cicero) [generated using https://github.com/elizabethdorans/E2G_Method_Tutorials]\n",
    "# File path to called peaks (macs2_peaks.bed)\n",
    "# Also define a list of dataset-cell type labels, which will be appending to feature names\n",
    "# Each of these lists should be of the same length, with one entry per set of input scores\n",
    "\n",
    "# Examples are provided and should be replaced with file paths of interest\n",
    "signac_files = [\"example_input_files/signac_peak_gene_links.tsv\"] ### REPLACE/ADD FILEPATHS\n",
    "scent_files = [\"example_input_files/scent_peak_gene_links.tsv\"] ### REPLACE/ADD FILEPATHS\n",
    "cicero_files = [\"example_input_files/cicero_peak_gene_links.tsv\"] ### REPLACE/ADD FILEPATHS\n",
    "peaks_files = [\"example_input_files/macs2_peaks.bed\"] ### REPLACE/ADD FILEPATHS\n",
    "dataset_celltype_labels = ['dataset1_celltype1'] ### REPLACE/ADD LABELS\n",
    "\n",
    "# Linking distance: maximum peak-TSS distance for candidate peak-gene links (default is TSS +/- 500kb)\n",
    "linking_distance_threshold = 500000 ## Recommended parameter\n",
    "\n",
    "# Promoter distance: maximum distance from a gene's TSS defined as the gene's promoter (default is TSS +/- 1kb)\n",
    "promoter_distance_threshold = 1000 ## Recommended parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a901f02f-60c3-48ba-83b2-37bcc15f3238",
   "metadata": {},
   "outputs": [],
   "source": [
    "## IMPORT PACKAGES\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pybedtools\n",
    "import os\n",
    "\n",
    "## LOAD RESOURCES\n",
    "\n",
    "# Load SNP coordinates (1kG SNPs with MAF count > 5)\n",
    "snps_1kg = pybedtools.BedTool(\"resources/SNPs.bed.gz\")\n",
    "\n",
    "# Load TSS coordinates\n",
    "tss = pd.read_csv(\"resources/TSS.txt.gz\", sep = '\\t')\n",
    "tss['chr'] = 'chr' + tss['chr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6baa33ff-f1df-424e-a2c3-4670cb1ca78e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## DEFINE FUNCTIONS \n",
    "\n",
    "# Adds a 'peak' column based on chromosome, start, and end columns\n",
    "def add_peak_column(df, chr_col = 'chr', start_col = 'start', end_col = 'end'):\n",
    "    df['peak'] = df[chr_col].astype(str) + '-' + df[start_col].astype(str) + '-' + df[end_col].astype(str)\n",
    "    return df\n",
    "\n",
    "# Adds chromosome, start, and end columns based on a 'peak' column\n",
    "def add_bed_columns(df, element_col = \"peak\"):\n",
    "    # adjust_start_coords = True when processing peak-gene links from Signac, SCENT, Cicero (start coordinate is +1 from peak coordinates called by macs2)\n",
    "    prev_cols = df.columns.to_list()\n",
    "    df[[\"chr\", \"start\", \"end\"]] = df[element_col].str.split(\"-\", expand = True)\n",
    "    df[[\"start\", \"end\"]] = df[[\"start\", \"end\"]].astype(int)\n",
    "    df = df[[\"chr\", \"start\", \"end\"] + prev_cols]\n",
    "    return df\n",
    "\n",
    "# Adjusts peak start coordinates (constituent methods output starting coordinates +1 relative to called peaks; these are adjusted to merge candidate links)\n",
    "def adjust_peak_start_coordinates(df, start_adjustment = -1):\n",
    "    df = add_bed_columns(df)\n",
    "    df['start'] = df['start'] - 1\n",
    "    df = add_peak_column(df)\n",
    "    return df.drop(columns = ['chr', 'start', 'end'])\n",
    "    \n",
    "# Read linking score predictions from constituent methods (generated using tutorial code in https://github.com/elizabethdorans/E2G_Method_Tutorials)\n",
    "def read_links(file, score_column, method):\n",
    "    links = pd.read_csv(file, sep = \"\\t\", comment='#')\n",
    "    links = adjust_peak_start_coordinates(links)\n",
    "    links = links.rename(columns = {score_column: method})\n",
    "    links = links[['peak', 'gene', method]]\n",
    "    return links\n",
    "\n",
    "# Define candidate peak-gene links by intersecting gene windows with called peaks\n",
    "def get_candidate_peak_gene_links(peaks_bedtool, linking_windows_bedtool):\n",
    "    candidate_pgl = peaks_bedtool.intersect(linking_windows, wa = True, wb = True).to_dataframe()\n",
    "    candidate_pgl = candidate_pgl.drop(\n",
    "        columns = [\"score\", \"strand\", \"thickStart\"])\n",
    "    candidate_pgl = candidate_pgl.rename(\n",
    "        columns = {\"chrom\": \"chr\", 'name': 'peak', \"thickEnd\": \"gene\"})\n",
    "    return candidate_pgl\n",
    "\n",
    "# Detect peaks that overlap promoters (these will be excluded from candidate links)\n",
    "def get_peaks_in_promoters(peaks_bedtool, promoter_windows_bedtool):\n",
    "    peaks_in_promoters = peaks_bedtool.intersect(promoter_windows, wa = True, wb = True).to_dataframe()['name'].tolist()\n",
    "    return peaks_in_promoters\n",
    "\n",
    "# Adds a DF column for distance between peak/SNP and gene\n",
    "def add_distance_to_gene(df, tss = tss, element_col = 'peak', gene_col = 'gene'):\n",
    "    if 'chr' not in df.columns:\n",
    "        df = add_bed_columns(df, element_col = element_col)\n",
    "    df = df.merge(tss, on = [gene_col, \"chr\"]).copy()\n",
    "    df['loc'] = np.ceil(df[['start', 'end']].mean(axis = 1))\n",
    "    df['%s_gene_distance' % element_col] = abs(df[\"loc\"] - df[\"tss\"]).astype(int)\n",
    "    return df.drop(columns = ['chr', 'start', 'end', 'loc', 'tss'])\n",
    "\n",
    "# Converts peak-gene links to SNP-gene links by intersecting peaks with SNPs\n",
    "def peak_gene_links_to_snp_gene_links(peak_gene_links, snp_bed):\n",
    "    pgl = peak_gene_links[[\"peak\", \"gene\"]].drop_duplicates()\n",
    "    pgl = add_bed_columns(pgl)\n",
    "    sgl = pybedtools.BedTool.from_dataframe(pgl[[\"chr\", \"start\", \"end\", \"peak\", \"gene\"]]).intersect(\n",
    "        snp_bed).to_dataframe().rename(columns = {\"score\": \"gene\", \"name\": \"peak\"})\n",
    "    if len(sgl) == 0:\n",
    "        sgl = pd.DataFrame(columns = ['snp', 'peak', 'gene'])\n",
    "    else:\n",
    "        sgl[\"snp\"] = sgl[\"chrom\"].astype(str) + \"-\" + sgl[\"start\"].astype(str) + \"-\" + sgl[\"end\"].astype(str)\n",
    "    sgl = sgl[[\"snp\", \"peak\", \"gene\"]].merge(peak_gene_links, how = \"left\", on = [\"peak\", \"gene\"])\n",
    "    return sgl\n",
    "\n",
    "# Filters peak-gene links by distance to TSS (< 500kb and >1kb, by default, see parameters defined above)\n",
    "def process_links(df, peaks_in_promoter, linking_distance_threshold):\n",
    "    # Restrict to candidate links within specified distance to TSS\n",
    "    df = add_distance_to_gene(df, element_col = 'peak')\n",
    "    df = df[df['peak_gene_distance'] < linking_distance_threshold]\n",
    "    # Remove candidate links with a peak overlapping a promoter\n",
    "    df = df[~df['peak'].isin(peaks_in_promoter)]\n",
    "    return df.drop(columns = 'peak_gene_distance')\n",
    "\n",
    "# Generates pgBoost features for a single dataset-cell type pair\n",
    "# Takes the following input files generated by tutorial code from https://github.com/elizabethdorans/E2G_Method_Tutorials: linking predictions from Signac, SCENT, Cicero & bedfile of ATAC peaks\n",
    "def create_features_single_dataset(signac_file, scent_file, cicero_file, peaks_file, snp_bed):\n",
    "    # Create Bedtool for peaks called by macs2\n",
    "    peaks = pd.read_csv(peaks_file, names = ['chr', 'start', 'end'], sep = \"\\t\", header = None, usecols = range(3))\n",
    "    peaks = add_peak_column(peaks)\n",
    "    peaks = pybedtools.BedTool.from_dataframe(peaks)\n",
    "\n",
    "    # Define peaks that overlap a promoter (these will be excluded from analyses) (default: TSS +/- 1kb)\n",
    "    peaks_in_promoters = get_peaks_in_promoters(peaks, promoter_windows)\n",
    "\n",
    "    # Define candidate peak-gene links (peaks within the specified distance to TSS, excluding peaks in promoters)\n",
    "    candidate_peak_gene_links = get_candidate_peak_gene_links(peaks, linking_windows)\n",
    "    candidate_peak_gene_links = process_links(candidate_peak_gene_links, peaks_in_promoters, linking_distance_threshold)\n",
    "    \n",
    "    print(\"%s candidate peak-gene links\" % len(candidate_peak_gene_links))\n",
    "    \n",
    "    # Define SNP-gene links by intersecting peak-gene links with SNPs\n",
    "    candidate_snp_gene_links = peak_gene_links_to_snp_gene_links(candidate_peak_gene_links, snps_1kg)\n",
    "    print(\"%s candidate SNP-gene links\" % len(candidate_snp_gene_links))\n",
    "\n",
    "    # Read linking scores from constituent methods\n",
    "    signac_links = read_links(signac_file, score_column = \"Score\", method = \"signac\")\n",
    "    signac_links = peak_gene_links_to_snp_gene_links(signac_links, snp_bed)\n",
    "    scent_links = read_links(scent_file,  score_column = \"beta\", method = \"scent\")\n",
    "    scent_links = peak_gene_links_to_snp_gene_links(scent_links, snp_bed)\n",
    "    cicero_links = read_links(cicero_file,  score_column = \"Score\", method = \"cicero\")\n",
    "    cicero_links = peak_gene_links_to_snp_gene_links(cicero_links, snp_bed)\n",
    "    \n",
    "    # Merge all constituent scores with candidate SNP-gene links\n",
    "    for links in [signac_links, scent_links, cicero_links]:\n",
    "        candidate_snp_gene_links = candidate_snp_gene_links.merge(links, on = ['snp', 'peak', 'gene'], how = 'left')\n",
    "    print(candidate_snp_gene_links)\n",
    "        \n",
    "    return candidate_snp_gene_links\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c35d19cb-c60e-409c-83da-984cc46ea8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "## DEFINE CANDIDATE LINKS\n",
    "\n",
    "# Define windows around TSS for candidate peak-gene links (e.g. <500kb)\n",
    "linking_windows = tss.copy()\n",
    "linking_windows[\"start\"] = (linking_windows[\"tss\"] - linking_distance_threshold).clip(0)\n",
    "linking_windows[\"end\"] = linking_windows[\"tss\"] + linking_distance_threshold\n",
    "linking_windows = linking_windows[[\"chr\", \"start\", \"end\", \"gene\"]]\n",
    "linking_windows = pybedtools.BedTool.from_dataframe(linking_windows)\n",
    "\n",
    "# Define windows around TSS for promoters (e.g. <1kb)\n",
    "promoter_windows = tss[['chr', 'tss', 'gene']].copy()\n",
    "promoter_windows['start'] = (tss['tss'] - promoter_distance_threshold).clip(0)\n",
    "promoter_windows['end'] = tss['tss'] + promoter_distance_threshold\n",
    "promoter_windows = promoter_windows[['chr', 'start', 'end', 'gene']]\n",
    "promoter_windows = pybedtools.BedTool.from_dataframe(promoter_windows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af18e78f-40f4-4786-a26f-d43648fc04c5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# For each dataset/cell type, read in all constituent scores and combine to form feature dataframe\n",
    "features = pd.DataFrame(columns = ['snp', 'gene'])\n",
    "if len(signac_files) > 0 and (len(signac_files) == len(scent_files) == len(cicero_files) == len(peaks_files) == len(dataset_celltype_labels)):\n",
    "    inputs = zip(signac_files, scent_files, cicero_files, peaks_files, dataset_celltype_labels)\n",
    "    for signac_file, scent_file, cicero_file, peaks_file, dataset_celltype_label in inputs:\n",
    "        print(\"\\n\", dataset_celltype_label)\n",
    "        candidate_sgl = create_features_single_dataset(signac_file, scent_file, cicero_file, peaks_file, snps_1kg)\n",
    "        renamed_columns = [\"%s_%s\" % (col, dataset_celltype_label) if col not in ['snp', 'gene'] else col for col in candidate_sgl.columns]\n",
    "        candidate_sgl.columns = renamed_columns\n",
    "        # Merge with features dataframe\n",
    "        features = features.merge(candidate_sgl, on = ['snp', 'gene'], how = 'outer')\n",
    "else:\n",
    "    print(\"Input lists must not be empty, and must be of equal length!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dc9fa131-b12f-4762-b593-a45e3e626027",
   "metadata": {},
   "outputs": [],
   "source": [
    "## ADD DISTANCE-BASED FEATURES\n",
    "\n",
    "# SNP-gene distance\n",
    "features = add_distance_to_gene(features, element_col = 'snp')\n",
    "\n",
    "# Closest TSS (binary)\n",
    "closest_tss = features.sort_values(by = [\"snp\", \"snp_gene_distance\"]\n",
    "                                 ).drop_duplicates(subset = \"snp\", keep = \"first\"\n",
    "                                 )[[\"snp\", \"gene\"]]\n",
    "closest_tss[\"closest_tss\"] = 1\n",
    "\n",
    "features = features.merge(closest_tss, how = \"left\", on = [\"snp\", \"gene\"])\n",
    "features[\"closest_tss\"] = features[\"closest_tss\"].fillna(0)\n",
    "features[\"closest_tss\"] = features[\"closest_tss\"].astype(int)\n",
    "\n",
    "# Add 'chr' column (for LOCO training)\n",
    "features['chr'] = features['snp'].str.split(\"-\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a197653-bd31-4860-b851-da1253602402",
   "metadata": {},
   "outputs": [],
   "source": [
    "## SAVE OUTPUT (feature matrix and column lists)\n",
    "\n",
    "# Classify columns\n",
    "drop_duplicates_columns = []\n",
    "feature_columns = []\n",
    "for column in features.columns:\n",
    "    if column in ['snp', 'gene', 'chr']:\n",
    "        continue\n",
    "    elif column.startswith(\"peak\"):\n",
    "        drop_duplicates_columns.append(column)\n",
    "    else:\n",
    "        feature_columns.append(column)\n",
    "\n",
    "if output_folder != None:\n",
    "    if not os.path.exists(output_folder):\n",
    "        os.makedirs(output_folder)\n",
    "    \n",
    "    print(\"Outputting %s features for %s candidate SNP-gene links!\" % (len(feature_columns), len(features)))       \n",
    "    \n",
    "    # Save feature matrix (--data_file argument to pgBoost.R)\n",
    "    features.to_csv(\"%s/data.tsv\" % output_folder, sep = '\\t', index = False)\n",
    "    \n",
    "    # Save features list (--predictor_file argument to pgBoost.R)\n",
    "    with open(\"%s/predictors.txt\" % output_folder, 'w') as features_file:\n",
    "        features_file.write(\"\\n\".join(feature_columns))\n",
    "        \n",
    "    # Save drop duplicates columns (--drop_duplicates_file argument to pgBoost.R)\n",
    "    with open(\"%s/drop_duplicates.txt\" % output_folder, 'w') as drop_duplicates_file:\n",
    "        drop_duplicates_file.write(\"\\n\".join(drop_duplicates_columns))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
