# pgBoost

pgBoost is an integrative modeling framework that trains a non-linear combination of single-cell multiome peak-gene linking scores and genomic distance-based features on fine-mapped eQTL data to predict regulatory SNP-gene links (see [Dorans et al. 2025 Nature Genetics](https://www.nature.com/articles/s41588-025-02220-3)).

The pgBoost framework requires single-cell mutiome (RNA-seq + ATAC-seq) data as input. The algorithm will output probabilistic scores for candidate SNP-gene links. See below for the steps to generate pgBoost features from single-cell multiome data and run pgBoost!

<div align="center">
<img src="https://github.com/elizabethdorans/pgBoost/blob/main/figs/overview_method.jpg" width=75%>
</div>


## Step 1: Generate constituent scores

Tutorial code to generate linking scores from constituent methods has been provided [here](https://github.com/elizabethdorans/E2G_Method_Tutorials/). To run pgBoost, scores must be generated for [Signac](https://github.com/elizabethdorans/E2G_Method_Tutorials/tree/main/Signac), [SCENT](https://github.com/elizabethdorans/E2G_Method_Tutorials/tree/main/SCENT), and [Cicero](https://github.com/elizabethdorans/E2G_Method_Tutorials/tree/main/Cicero). 

<ins>NOTE</ins>: 
- When running SCENT, follow all instructions marked \*** to skip the p-value calculation (not needed for pgBoost input). This will save time and computational resources.
- _Do not_ perform the optional "post-processing for IGVF portal" step for any of the consituent score files.

## Step 2: Create pgBoost input files

After constituent scores have been generated, `prepare_pgBoost_features.ipynb` can be used to generate pgBoost input files. This notebook will generate distance-based features and output the --data_file, --predictor_file, and --drop_duplicates_file inputs needed for the pgBoost script (see below).

## pgBoost.R [![DOI](https://zenodo.org/badge/788634518.svg)](https://doi.org/10.5281/zenodo.15276307)

pgBoost takes as input a data set of candidate SNP-gene regulatory links x link attributes (peak-gene correlation-based scores from constituent linking methods, distance-based features) and generates consensus linking scores using gradient boosting (in a leave-one-chromosome-out framework).

### Arguments

Rscript pgBoost.R --data_file data.tsv --training_file `resources/training_data.tsv.gz` --predictor_file predictors.txt --drop_duplicates_file drop_duplicates.txt

Example input files are provided in the folder `example_input_files/`.

| Argument | Description |
| -------- | ----------- |
| __--data_file__ | A tab-separated data frame of candidate links (rows) x linking attributes (columns). Must contain all columns specified in __predictor_file__ _and_ the column specified in __LOO_colname__. Must also contain one or more columns which uniquely index candidate links (e.g. "SNP", "peak", "gene"). Additional columns will be ignored. `prepare_pgBoost_features.ipynb` will generate `data.tsv`. |
| __--training_file__ | A tab-separated data frame of training links (rows) x training link attributes (columns). Must contain one column named "positive" which provides a binary indicator (0/1) of whether a link is a positive (1) or negative (0) training instance. Must contain columns which uniquely identify candidate links (e.g. "SNP", "peak", "gene") matching those in __data_file__. The pgBoost training set has been provided at `resources/training_data.tsv.gz`. |
| __--predictor_file__ | A line-delimited text file containing names of columns to use as predictors. Must match columns in __data_file__. `prepare_pgBoost_features.ipynb` will generate `predictors.txt`. |
| __--drop_duplicates_file__ | (OPTIONAL) A line-delimited text file containing the names of columns used to drop duplicate instances during training. If more than one training instance contains the same values across these columns (and the same classification as positive/negative), only one instance will be retained (for example, if candidate links are SNP-gene pairs, we recommend using any 'peak' columns to drop duplicate instances, so that the same peak-gene link is not overused in training). Must match columns in __data_file__. If not supplied, pgBoost will not check for duplicate training instances. `prepare_pgBoost_features.ipynb` will generate `drop_duplicates.txt`. |
| __--LOO_colname__ | Name of the column used to group links for the leave-one-chromosome-out framework, e.g. "chr" (default), "CHR", "chrom", "chromosome". `prepare_pgBoost_features.ipynb` will generate a "chr" column in `data.tsv`. |
| __--outfile__ | Name of tab-delimited file where pgBoost predictions will be saved (default "pgboost_predictions.tsv"). |

### Output

The pgBoost output file will include:
- All index columns uniquely identifying candidate links (e.g. "SNP", "peak", "gene").
- __pgBoost_probability__: a column containing pgBoost predictions (probabilities generated by the gradient boosting classification algorithm).
- __pgBoost_percentile__: a column containing percentile rank by pgBoost_probability (we recommend a threshold of pgBoost_percentile >0.95 to define true links; see [Dorans et al. Nature Genetics 2025](https://www.nature.com/articles/s41588-025-02220-3)).

An example output file is provided in the folder `example_output_files/`.
